pysages/colvars/angles.py

```diff
@@ -121,3 +127,152 @@ def dihedral_angle(p1, p2, p3, p4):
+    by `1025` degree/nanometer.
+
+    Usage
+    -------
+    cvs = [PhaseAngle([index1, index2, ...])]
+
+    Notice that the phase angle is dependent on the order of the indices. For example,
+    the convention for sugar pucker of ribose in RNA/DNA is: O4', C1', C2', C3', C4'.
+    """
+
+    @property
+    def function(self):
+        return ring_puckering_coordinates
+
+
+def ring_puckering_coordinates(rs):
+    r"""
+    calculate phase angle (first phase angle if N>5) based on Cremer-Pople method.
+    :math:`r0 = 1/N \sum_i^N \vec{r}_i`
+    :math:`\vec{R1} = \sum_i^N (\vec{r}_i -r_c) \sin(2\pi (i-1)/N)`
+    :math:`\vec{R2} = \sum_i^N (\vec{r}_i -r_c) \cos(2\pi (i-1)/N)`
+    :math:`\hat{n} = \vec{R1} \times \vec{R2}/ (|\vec{R1}\times\vec{R2}|)`
+    :math:`z_i = (\vec_{r}_i-r_c) \cdot \hat{n}`
+    :math:`a =  \sqrt(2/N) \sum_i^N z_i \cos(2\pi 2(i-1)/N)`
+    :math:`b = -\sqrt(2/N) \sum_i^N z_i \sin(2\pi 2(i-1)/N)`
+    :math:`q = \sqrt(a^2 + b^2)`
+    :math:`phi = \atan(b / a)`
+
+    Parameters
+    ------------
+    rs: DeviceArray
+        :math: `\vec{r}_i` array of 3D vector in space
+
+    Returns
+    ------------
+    DeviceArray: [q: float, phi: float]
+        :math:`q` in nanometer (if the default length unit for the MD engine is nanometer)
+        :math:`phi` in radians, range -pi to pi.
+    """
+    N = len(rs)
+    r0 = barycenter(rs)
+    rc = rs - r0
+    theta = 2j * np.pi * np.arange(N) / N
+    fourier_coeff = np.exp(theta)
+    R1 = np.dot(rc.T, np.imag(fourier_coeff))
+    # Notice the imag part corresponds to sin. The order of R1/R2 matters
+    # because otherwise the n would be inverted.
+    R2 = np.dot(rc.T, np.real(fourier_coeff))
+    n = np.cross(R1, R2)
+    n /= linalg.norm(n)
+    z = np.dot(rc, n)
+    fourier_coeff2 = np.exp(2 * theta)
+    a = np.sqrt(2 / N) * np.sum(z * np.real(fourier_coeff2))
+    b = -np.sqrt(2 / N) * np.sum(z * np.imag(fourier_coeff2))
+    q = np.sqrt(a**2 + b**2)
+    phi = np.arctan2(b, a)
+    return np.array([q, phi])
+
+
+class RingPhaseAngle(CollectiveVariable):
+    """
+    Computes the phase angle of a monocyclic ring by the Cremer-Pople method.
+    Mathematical definitions can be found in
+    [D. Cremer and J. A. Pople, JACS, 1974](https://pubs.acs.org/doi/10.1021/ja00839a011)
+    Equations 4-14.
+    """
+
+    @property
+    def function(self):
+        """
+        Returns
+        -------
+        Function that calculates the dihedral angle value from a simulation snapshot.
+        Look at `pysages.colvars.angles.ring_puckering_coordinates`
+        and `pysages.colvars.angles.ring_phase_angle` for details.
+        """
+        return ring_phase_angle
+
+
+def ring_phase_angle(rs):
+    r"""
+    Parameters
+    ------------
+    rs: DeviceArray
+        :math: `\vec{r}_i` array of 3D vector in space
+
+    Returns
+    ------------
+    float
+        :math:`P` in range -pi to pi.
+    """
+    _, phi = ring_puckering_coordinates(rs)
+    return phi
```

Comment:
```suggestion
    a, b = ring_puckering_coordinates(rs)
    return np.arctan2(b, a)
```

Reply:
I see, but there might be a need for people to get both the q and the phi?

Reply:
That's fair. Not sure what would the most common scenario would be. Do you have any opinions on this @ricalessandri.

Reply:
I think the most useful one is the phase angle with the real info, the amplitude, at least in the 5-membered ring in RNA, is usually calculated with the phase angle. I haven't seen people who only calculate the amplitude, because usually, it fluctuates around some eq value, but phase angle can switch between different conformations.

Reply:
Ok. Let's leave as you have it now, and we will revisit if/when there's a demand for something more flexible.

Reply:
cool, thanks for the review! I'll try to add the test that Ludwig suggested today!